File: example/augas/sched.cpp
/*
  Copyright (c) 2004, 2005, 2006, 2007, 2008, 2009 Mark Aylett <mark.aylett@gmail.com>

  This file is part of Aug written by Mark Aylett.

  Aug is released under the GPL with the additional exemption that compiling,
  linking, and/or using OpenSSL is allowed.

  Aug is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation; either version 2 of the License, or (at your option) any later
  version.

  Aug is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
#define MOD_BUILD
#include "augmodpp.hpp"

#include "augnetpp.hpp"
#include "augutilpp.hpp"
#include "augmarpp.hpp"
#include "augsyspp.hpp"

#include <map>
#include <sstream>

using namespace aug;
using namespace mod;
using namespace std;

namespace {

    enum tmtz { TMLOCAL, TMUTC };

    struct tmevent : mpool_ops {
        const mod_id id_;
        const string name_, spec_;
        const tmtz tz_;
        aug_tmspec tmspec_;
        tmevent(mod_id id, const string& name, const string& spec, tmtz tz)
            : id_(id),
              name_(name),
              spec_(spec),
              tz_(tz)
        {
        }
        tmevent(const string& name, const string& spec, tmtz tz)
            : id_(aug_nextid()),
              name_(name),
              spec_(spec),
              tz_(tz)
        {
        }
    };

    typedef smartptr<tmevent> tmeventptr;
    typedef multimap<time_t, tmeventptr> tmqueue;

    string
    tmstring(const tm& local)
    {
        char buf[256];
        strftime(buf, sizeof(buf), "%b %d %H:%M:%S", &local);
        return buf;
    }

    void
    pushevent(tmqueue& q, time_t now, const tmeventptr& ptr)
    {
        tm tm;
        if (TMUTC == ptr->tz_) {

            if (!aug_nexttime(aug_gmtime(&now, &tm), &ptr->tmspec_))
                return;

            q.insert(make_pair(now = aug_timegm(&tm), ptr));

        } else {

            if (!aug_nexttime(aug_localtime(&now, &tm), &ptr->tmspec_))
                return;

            q.insert(make_pair(now = aug_timelocal(&tm), ptr));
        }

        aug_ctxinfo(aug_tlx, "event [%s] scheduled for %s",
                    ptr->name_.c_str(),
                    tmstring(*aug_localtime(&now, &tm)).c_str());
    }

    void
    pushevent(tmqueue& q, time_t now, const string& name, tmtz tz)
    {
        const char* tmspecs
            (mod::getenv(string("session.sched.event.").append(name)
                          .append(TMUTC == tz ? ".utc" : ".local")
                          .c_str()));
        if (!tmspecs)
            return;

        istringstream is(tmspecs);
        string spec;
        while (is >> spec) {

            tmeventptr ptr(new (tlx) tmevent(name, spec, tz));
            if (aug_strtmspec(&ptr->tmspec_, spec.c_str()))
                pushevent(q, now, ptr);
        }
    }

    void
    pushevents(tmqueue& q, time_t now)
    {
        const char* events(mod::getenv("session.sched.events"));
        if (!events)
            return;

        istringstream is(events);
        string name;
        while (is >> name) {
            pushevent(q, now, name, TMLOCAL);
            pushevent(q, now, name, TMUTC);
        }
    }

    void
    eraseevent(tmqueue& q, mod_id id)
    {
        tmqueue::iterator it(q.begin()), end(q.end());
        for (; it != end; ++it)
            if (it->second->id_ == id) {
                q.erase(it);
                break;
            }
    }

    unsigned
    timerms(const tmqueue& q, const timeval& tv)
    {
        if (q.empty())
            return 0;

        tmqueue::const_iterator next(q.begin());
        timeval expiry = { next->first, 0 };
        tvsub(expiry, tv);
        unsigned ms(tvtoms(expiry));
        return ms < 60000 ? ms : 60000;
    }

    ostream&
    toxml(ostream& os, const tmqueue& q, unsigned offset, unsigned max_)
    {
        offset = AUG_MIN(offset, q.size());

        os << "<result><events offset=\"" << offset
           << "\" total=\"" << static_cast<unsigned>(q.size()) << "\">";
        tmqueue::const_iterator it(q.begin()), end(q.end());
        for (advance(it, offset); it != end && max_; ++it, --max_) {
            tm tm;
            os << "<event id=\"" << it->second->id_
               << "\" name=\"" << it->second->name_
               << "\" spec=\"" << it->second->spec_ << "\" tz=\""
               << (TMUTC == it->second->tz_ ? "utc" : "local")
               << "\">" << tmstring(*aug_localtime(&it->first, &tm))
               << "</event>";
        }
        os << "</events></result>";
        return os;
    }

    struct sched : basic_session, mpool_ops {
        mod_id timer_;
        tmqueue queue_;
        void
        checkexpired(const timeval& tv)
        {
            time_t now(tv.tv_sec);

            while (!queue_.empty() && queue_.begin()->first <= now) {

                tmeventptr ptr(queue_.begin()->second);
                queue_.erase(queue_.begin());

                // Inform interested modules of expired event.

                mod_post(0, "sched-client", ptr->name_.c_str(), 0);
                pushevent(queue_, now, ptr);
            }
        }
        void
        settimer(const timeval& tv)
        {
            checkexpired(tv);

            unsigned ms(timerms(queue_, tv));
            if (ms) {
                if (timer_)
                    resettimer(timer_, ms);
                else
                    timer_ = mod::settimer(ms, 0);
                aug_ctxinfo(aug_tlx, "next expiry in %d ms", ms);
            } else if (timer_) {
                canceltimer(timer_);
                timer_ = 0;
            }
        }
        void
        delevent(const map<string, string>& params)
        {
            mod_id id(getvalue<mod_id>(params, "id"));

            aug_ctxinfo(aug_tlx, "deleting event: id=[%u]", id);
            eraseevent(queue_, id);

            timeval tv;
            gettimeofday(tv);
            settimer(tv);
        }
        void
        putevent(const map<string, string>& params)
        {
            mod_id id(getvalue<mod_id>(params, "id"));

            if (id) {
                aug_ctxinfo(aug_tlx, "updating event: id=[%u]", id);
                eraseevent(queue_, id);
            } else {
                aug_ctxinfo(aug_tlx, "inserting new event");
                id = aug_nextid();
            }

            string name(getvalue<string>(params, "name"));
            string spec(getvalue<string>(params, "spec"));

            tmtz tz(TMUTC);
            if (getvalue<string>(params, "tz") == "local")
                tz = TMLOCAL;

            timeval tv;
            gettimeofday(tv);

            tmeventptr ptr(new (tlx) tmevent(id, name, spec, tz));
            if (aug_strtmspec(&ptr->tmspec_, spec.c_str()))
                pushevent(queue_, tv.tv_sec, ptr);

            settimer(tv);
        }
        void
        respond(mod_id id, const char* from, const char* type, blobref blob)
        {
            size_t size;
            const char* content(static_cast<const char*>
                                (aug::getblobdata(blob, size)));

            map<string, string> params;
            unpackurl(content, content + size,
                      inserter(params, params.begin()));

            marptr mar(createmar(getmpool(aug_tlx)));
            putfieldp(mar, "Cache-Control", "no-cache");
            putfieldp(mar, "Content-Type", "text/xml");

            if (0 == strcmp(type, "service/reconf")) {

                reconfall();
                setcontent(mar, "<message type=\"info\">"
                           "re-configured</message>");
            } else {

                if (0 == strcmp(type, "service/sched/delevent")) {
                    delevent(params);
                } else if (0 == strcmp(type, "service/sched/putevent")) {
                    putevent(params);
                } else if (0 != strcmp(type, "service/sched/events")) {
                    aug_ctxwarn(aug_tlx, "event [%s] ignored", type);
                    return;
                }

                unsigned offset(getvalue<unsigned>(params, "offset"));
                unsigned max_(getvalue<unsigned>(params, "max"));

                omarstream os(mar);
                toxml(os, queue_, offset, max_);
            }

            dispatch(id, from, "result", mar.base());
        }
        bool
        do_start(const char* sname)
        {
            do_reconf();
            return true;
        }
        void
        do_reconf()
        {
            timeval tv;
            gettimeofday(tv);

            queue_.clear();
            pushevents(queue_, tv.tv_sec);
            settimer(tv);
        }
        void
        do_event(mod_id id, const char* from, const char* type,
                 aug_object_* ob)
        {
            aug_ctxinfo(aug_tlx, "event [%s] triggered", type);

            marptr mar(object_cast<aug_mar>(obptr(ob)));
            if (null == mar)
                return;

            const void* value;
            getfieldp(mar, "Content-Type", value);
            if (!value || 0 != strcmp(static_cast<const char*>(value),
                                      "application/x-www-form-urlencoded"))
                return;

            respond(id, from, type, object_cast<aug_blob>(mar));
        }
        void
        do_expire(const handle& timer, unsigned& ms)
        {
            timeval tv;
            gettimeofday(tv);

            checkexpired(tv);

            ms = timerms(queue_, tv);
            aug_ctxinfo(aug_tlx, "next expiry in %d ms", ms);
        }
        ~sched() AUG_NOTHROW
        {
        }
        sched()
            : timer_(0)
        {
        }
        static session_base*
        create(const char* sname)
        {
            return new (tlx) sched();
        }
    };

    typedef basic_module<basic_factory<sched> > module;
}

MOD_ENTRYPOINTS(module::init, module::term)

File: example/augas/htdocs/sched.js
// -*- java -*-

/*
  Copyright (c) 2004, 2005, 2006, 2007, 2008, 2009 Mark Aylett <mark.aylett@gmail.com>

  This file is part of Aug written by Mark Aylett.

  Aug is released under the GPL with the additional exemption that compiling,
  linking, and/or using OpenSSL is allowed.

  Aug is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation; either version 2 of the License, or (at your option) any later
  version.

  Aug is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51
  Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

function Event(id, name, spec, tz, next) {
    this.id = id;
    this.name = name;
    this.spec = spec;
    this.tz = tz;
    this.next = next;
}

function getEventParser(setEvents) {

    return function(xml) {

        var offset = xml.documentElement.getAttribute('offset');
        var total = xml.documentElement.getAttribute('total');
        var items = [];

        iterate(function(i, x) {

                var id = x.getAttribute('id');
                var name = x.getAttribute('name');
                var spec = x.getAttribute('spec');
                var tz = x.getAttribute('tz');
                var next = x.childNodes[0].nodeValue;
                items.push(new Event(id, name, spec, tz, next));

            }, xml.getElementsByTagName('event'));

        if (offset == null)
            offset = 0;

        if (total == null)
            total = items.length;

        setEvents(items, offset, total);
    };
}

function Events(div, log) {

    var pager_ = null;

    var parseXml = function(xml) {

        var offset = xml.documentElement.getAttribute('offset');
        var total = xml.documentElement.getAttribute('total');
        var items = [];

        iterate(function(i, x) {

                var id = x.getAttribute('id');
                var name = x.getAttribute('name');
                var spec = x.getAttribute('spec');
                var tz = x.getAttribute('tz');
                var next = x.childNodes[0].nodeValue;
                items.push(new Event(id, name, spec, tz, next));

            }, xml.getElementsByTagName('event'));

        if (offset == null)
            offset = 0;

        if (total == null)
            total = items.length;

        pager_.setItems(items, offset, total);
    };

    var refreshTable = function(items, offset, total) {

        var current = pager_.getCurrentId();

        var html = '<table width="100%">'
        + '<tr class="header"><th align="left">name</th>'
        + '<th align="left">spec</th><th align="left">tz</th>'
        + '<th align="left">next</th></tr>';

        iterate(function(i, x) {

                if (x.id == current) {
                    html += '<tr class="current">';
                } else {
                    html += '<tr class="item" onclick="setCurrent('
                        + x.id + ')">';
                }
                html += '<td>' + x.name + '</td>';
                html += '<td>' + x.spec + '</td>';
                html += '<td>' + x.tz + '</td>';
                html += '<td>' + x.next + '</td></tr>';
            }, items);

        for (var i = items.length; i < 5; ++i)
            html += '<tr class="empty"><td colspan="4">&nbsp;</td></tr>';

        html += '</table>';

        var prev = pager_.getPrevId(current);
        var next = pager_.getNextId(current);

        html += '<p>';
        html += '<input class="action" type="button" value="prev"'
        if (prev)
            html += ' onclick="setCurrent(' + prev + ');"/>';
        else
            html += ' disabled="true"/>';

        html += '<input class="action" type="button" value="next"'
        if (next)
            html += ' onclick="setCurrent(' + next + ');"/>';
        else
            html += ' disabled="true"/>';

        html += '<input class="action" type="button" value="delete"';
        if (current)
            html += ' onclick="delEvent(' + current + ');"/>';
        else
            html += ' disabled="true"/>';
        html += '</p>';

        div.innerHTML = html;
    };

    var refreshForm = function(items, offset, total) {

        var current = pager_.getCurrentId();

        var x = null;
        if (current && (x = getById(items, current)))
            x = x.value;
        else {
            current = 0;
            x = new Event('', '', '', 'local', '');
        }

        document.getElementById('name').value = x.name;
        document.getElementById('spec').value = x.spec;
        document.getElementById('tz').value = x.tz;
    };

    var refresh = function(items, offset, total) {
        refreshTable(items, offset, total);
        refreshForm(items, offset, total);
    };

    var request = function(offset, max) {
        getXml(parseXml, 'service/sched/events?'
               + encodePairs([new Pair('offset', offset),
                              new Pair('max', max)]));
    };

    this.loadEvents = function() {
        log.add('info', 'load events');
        request(0, 5);
    };

    this.delEvent = function(id) {
        log.add('info', 'del event: ' + id);
        pager_.setNearId(id);
        getXml(parseXml, 'service/sched/delevent?'
               + pager_.encodePairs([new Pair('id', id)]));
    };

    this.putEvent = function() {
        log.add('info', 'put event');
        getXml(parseXml, 'service/sched/putevent?'
               + pager_.encodePairs([new Pair('id', pager_.getCurrentId()),
                                     getPairById('name'),
                                     getPairById('spec'),
                                     getPairById('tz')]));
    };

    this.setCurrent = function(id) {
        log.add('info', 'set current: ' + id);
        pager_.setCurrentId(id);
    };

    pager_ = new Pager(refresh, request, 5);
}

var log = null;
var events = null;
var loadEvents = null;
var delEvent = null;
var putEvent = null;
var setCurrent = null;

function init() {
    log = new Log(document.getElementById('log'));
    events = new Events(document.getElementById('view'), log);
    loadEvents = events.loadEvents;
    delEvent = events.delEvent;
    putEvent = events.putEvent;
    setCurrent = events.setCurrent;
    loadEvents();
}

function reconf() {
    log.add('info', 'reconf');
    getXml(log.addXml, 'service/reconf');
}
